contracts/Staking.sol

```diff
@@ -0,0 +1,488 @@
+pragma solidity 0.6.4;
+
+import "./System.sol";
+import "./lib/Memory.sol";
+import "./lib/BytesToTypes.sol";
+import "./interface/IApplication.sol";
+import "./interface/ICrossChain.sol";
+import "./interface/IParamSubscriber.sol";
+import "./interface/ITokenHub.sol";
+import "./lib/CmnPkg.sol";
+import "./lib/SafeMath.sol";
+import "./lib/RLPEncode.sol";
+import "./lib/RLPDecode.sol";
+import "./interface/IStaking.sol";
+
+contract Staking is IStaking, System, IParamSubscriber, IApplication {
+  using SafeMath for uint256;
+  using RLPEncode for *;
+  using RLPDecode for *;
+
+  struct DelegateAckPackage {
+    address delegator;
+    address validator;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct UndelegateAckPackage {
+    address delegator;
+    address validator;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct RedelegateAckPackage {
+    address delegator;
+    address valSrc;
+    address valDst;
+    uint256 amount;
+    uint8 errCode;
+  }
+
+  struct TransferInRewardSynPackage {
+    uint256[] amounts;
+    address[] recipients;
+    address[] refundAddrs;
+  }
+
+  struct TransferInUndelegatedSynPackage {
+    uint256 amount;
+    address recipient;
+    address refundAddr;
+    address validator;
+  }
+
+  // Cross-Chain Stake Event type
+  uint8 public constant EVENT_DELEGATE = 0x01;
+  uint8 public constant EVENT_UNDELEGATE = 0x02;
+  uint8 public constant EVENT_REDELEGATE = 0x03;
+  uint8 public constant EVENT_TRANSFER_IN_REWARD = 0x04;
+  uint8 public constant EVENT_TRANSFER_IN_UNDELEGATED = 0x05;
+
+  uint32 public constant ERROR_UNKNOWN_PACKAGE_TYPE = 101;
+  uint32 public constant ERROR_WITHDRAW_BNB = 102;
+
+  uint256 constant public TEN_DECIMALS = 1e10;
+
+  uint256 public constant INIT_ORACLE_RELAYER_FEE = 2e16; //TODO
+  uint256 public constant INIT_MIN_DELEGATION_CHANGE = 1e19;
+  uint256 public constant INIT_CALLBACK_GAS_LIMIT = 23000;
+
+  uint256 public oracleRelayerFee;
+  uint256 public minDelegationChange;
+
+  mapping(address => uint256) delegated;
+  mapping(address => mapping(address => uint256)) delegatedOfValidator;
+  mapping(address => uint256) distributedReward;
+  mapping(address => mapping(address => uint256)) pendingUndelegated;
+  mapping(address => uint256) undelegated;
+
+  bool internal locked;
+
+  modifier noReentrant() {
+    require(!locked, "No re-entrancy");
+    locked = true;
+    _;
+    locked = false;
+  }
+
+  modifier tenDecimalPrecision(uint256 amount) {
+    require(msg.value%TEN_DECIMALS==0,
```

Comment:
I think `tenDecimalPrecision` is a bit redundant, only need `require(amount%TEN_DECIMALS==0)`? If we need check msg.value, we input it as `amount`